flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -208,13 +208,28 @@ convertOptExtentExpr(Fortran::lower::AbstractConverter &converter,
 /// Does this expr designate an allocatable or pointer entity ?
 static bool
 isAllocatableOrPointer(const Fortran::lower::SomeExpr &expr,
-                       Fortran::lower::AbstractConverter &converter) {
-  // First check if this is a pointer function result.
-  if (Fortran::evaluate::IsObjectPointer(expr, converter.getFoldingContext()))
-    return true;
-  const Fortran::semantics::Symbol *sym =
-      Fortran::evaluate::UnwrapWholeSymbolOrComponentDataRef(expr);
-  return sym && Fortran::semantics::IsAllocatableOrPointer(*sym);
+                       Fortran::evaluate::FoldingContext &foldingContext) {
+  // Ensure pointer function results are considered to be pointers here.
+  return Fortran::evaluate::IsAllocatableOrPointerObject(expr, foldingContext);
+}
+
+static bool
+mayBePassedAsAbsentOptional(const Fortran::lower::SomeExpr &expr,
+                            Fortran::evaluate::FoldingContext &foldingContex) {
+  return Fortran::evaluate::MayBePassedAsAbsentOptional(expr, foldingContex);
```

Comment:
This function and the isAllocatableOrPointer() above don't do anything at this point. Just call `Fortran::evaluate::MayBePassedAsAbsentOptional()` (and `Fortran::evaluate::IsAllocatableOrPointerObject()`) directly.